cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java

```diff
@@ -201,11 +217,54 @@ private void updateWorkflowToken(WorkflowToken token, Map<String, StageStatistic
   }
 
   @Override
+  @SuppressWarnings("unchecked")
   protected SparkCollection<Object> handleAutoJoin(String stageName, JoinDefinition joinDefinition,
-                                                 Map<String, SparkCollection<Object>> inputDataCollections,
-                                                 @Nullable Integer numPartitions) {
-    //TODO: Add logic to execute join operation on SQL Engine.
+                                                   Map<String, SparkCollection<Object>> inputDataCollections,
+                                                   @Nullable Integer numPartitions) {
+
+    if (canJoinOnSQLEngine(joinDefinition, inputDataCollections)) {
+      SQLEngineJob<SQLDataset<Object>> job = sqlEngineAdapter.join(stageName,
+                                                                   joinDefinition,
+                                                                   inputDataCollections);
+      return new RDDCollection<Object>(sec, functionCacheFactory, jsc,
```

Comment:
it seems like it would be clearer if we used the adapter here to pull the result and create a normal RDDCollection rather than have some hybrid logic within RDDCollection.

Reply:
I assume we'll also have some performance optimizations later to avoid pulling if the join result is only ever used as the input to another SQL join.

Reply:
Yes, if the dataset already exists in the engine (because it was pushed before, or it was generated by the engine) we don't invoke this push operation as it is not necessary. https://github.com/cdapio/cdap/pull/13343/files#diff-99c96a996115d1c6721f78ffe7a41d2ddb7dae80420d24cf51ffa719b7c448deR247-R252

Reply:
@albertshau the argument against pulling here is that the output from this job might not be needed.

If I have a pipeline with 2 joins that then are used in a third join, pulling after the original 2 joins is unnecessary.

The rationale behind creating an RDD collection backed by a SQL Job is that we can wait for this job and pull if and only if we need to.

Reply:
Ok I see. Then I think it's cleaner to have a separate "SparkCollection" implementation that is purely backed by the SQLJob. Whenever any non-join method is called, it pulls, creates an RDDCollection, then delegates the call to the RDDCollection.

That way it's more explicit what is happening, instead of tying the pulling to the getUnderlying() method and having RDDCollection sometimes not have an RDD.

Reply:
I have implemented a `SQLEngineCollection` that delegates to the RDD collection if needed, but executes joins in the SQL engine.

Note that, as a pre-requisite, all collections must have been pushed previously (which we handle in the `handleAutoJoin` method of the pipeline runner.